File: Algorithms/C++/00203-Remove Linked List Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* result = head;
        ListNode* prev = nullptr;

        while(head != nullptr){
            if(head->val == val){
                if(prev == nullptr){
                    head = head->next;
                    result = head;
                } else if(head->next == nullptr){
                    prev->next = nullptr;
                    prev = head;
                    head = head->next;
                } else{
                    while(head->next != nullptr && head->next->val == val){
                        head = head->next;
                    }
                    prev->next = head->next;
                    prev = head;
                    head = head->next;
                }
                
            } else{
                prev = head;
                head = head->next;
            }

        }

        return result;
    }
};

/*
    - have a prev node
    case 1: target node at start
    case 2: target node in middle
    case 3: target node at end

*/

File: Algorithms/Python3/00203-Remove Linked List Elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        prev = None
        curr = head

        # move curr to first non-target node
        while curr and curr.val == val:
            curr = curr.next
        
        start = curr
        # iterate through list, skipping target nodes
        while curr:
            # search for non-target node
            while curr.val == val:
                curr = curr.next
                prev.next = curr
                if not curr:
                    return start     
            prev = curr
            curr = curr.next

        return start      
        
'''
    track prev node
    
    iterate through list to check vals
    if we see target, set prev node to next node
    while next node is not target node (or end of list)
    
    1 - 6 - 6 - 6 - 1

    alternatively:
        - use extra memory to create a new list with non-target nodes

'''